src/Microsoft.AspNetCore.Mvc.Core/Builder/MvcEndpointInfo.cs

```diff
@@ -5,106 +5,101 @@
+                    string.Format(CultureInfo.CurrentCulture, "An error occurred while creating the route with name '{0}' and pattern '{1}'.", name, pattern), exception);
             }
         }
 
-        public string Name { get; }
-        public string Template { get; }
-
-        // Non-inline defaults
-        public RouteValueDictionary Defaults { get; }
-
-        // Inline and non-inline defaults merged into one
-        public RouteValueDictionary MergedDefaults { get; }
-
-        public IDictionary<string, IRouteConstraint> Constraints { get; }
-        public RouteValueDictionary DataTokens { get; }
-        internal RouteTemplate ParsedTemplate { get; private set; }
-
-        private static IDictionary<string, IRouteConstraint> GetConstraints(
-            IInlineConstraintResolver inlineConstraintResolver,
-            RouteTemplate parsedTemplate,
-            IDictionary<string, object> constraints)
+        private Dictionary<string, IList<IRouteConstraint>> BuildConstraints(ParameterPolicyFactory parameterPolicyFactory)
         {
-            var constraintBuilder = new RouteConstraintBuilder(inlineConstraintResolver, parsedTemplate.TemplateText);
-
-            if (constraints != null)
-            {
-                foreach (var kvp in constraints)
-                {
-                    constraintBuilder.AddConstraint(kvp.Key, kvp.Value);
-                }
-            }
+            var constraints = new Dictionary<string, IList<IRouteConstraint>>(StringComparer.OrdinalIgnoreCase);
 
-            foreach (var parameter in parsedTemplate.Parameters)
+            foreach (var parameter in ParsedPattern.Parameters)
             {
-                if (parameter.IsOptional)
+                foreach (var parameterPolicy in parameter.ParameterPolicies)
                 {
-                    constraintBuilder.SetOptional(parameter.Name);
-                }
+                    var createdPolicy = parameterPolicyFactory.Create(parameter, parameterPolicy);
+                    if (createdPolicy is IRouteConstraint routeConstraint)
+                    {
+                        if (!constraints.TryGetValue(parameter.Name, out var paramConstraints))
+                        {
+                            paramConstraints = new List<IRouteConstraint>();
+                            constraints.Add(parameter.Name, paramConstraints);
+                        }
 
-                foreach (var inlineConstraint in parameter.InlineConstraints)
-                {
-                    constraintBuilder.AddResolvedConstraint(parameter.Name, inlineConstraint.Constraint);
+                        paramConstraints.Add(routeConstraint);
+                    }
                 }
             }
 
-            return constraintBuilder.Build();
+            return constraints;
```

Comment:
If this a placeholder for parameter transformers? We don't really need this for constraints, since routepattern will do this.

Reply:
RoutePattern will parse inline policies and build a RouteParameterPolicyReference with that string content. But that content needs to be resolved to an IRouteConstraint using the type map from RouteOptions. That is what is going on here.

Reply:
Ah right, I remember, it's because you execute the constraints 👍 I read this far and commented without thinking like a sillypants!